### DISCUSSION TITLE
What is the recommended way to implement "database per tenant" for a multitenant application?

### DISCUSSION THREAD

**Discussion by sadeghhosseini** [Automattic/mongoose#13265] [Q&A] [answered]
For a "Database per tenant" multitenancy approach. I could find two solutions:
1. creating a connection and using "[[Connection.prototype.useDb()](https://mongoosejs.com/docs/api/connection.html#Connection.prototype.useDb())]
2. creating a connection per tenant and caching the connection objects on a javascript map

Which one of these approaches is the recommended approach?

@vkarpov15

**Comment by Uzlopak** [comment]
I doubt somehow that you have a mongoserver per database.

The Mongo Driver itself already uses a connection pool. You probably hit then the maximum amount of connections when using multiple connection pools. So if you want to use multiple databases, then you probably should use useDB and switch the database every time.

On the other hand, it makes maybe more sense to not separate the data on database level but on collection level. So every tenant has its own specific collections.

Also it makes maybe more sense to already think about sharding. E.g. shard by tenant id. If you separate the tenant data by database or collection, than maybe then you probably already have no way to use sharding in a purposeful manner.

**Reply by sadeghhosseini** [reply]
Thank you for taking the time to answer my question.
I was using this [article by mongodb.com](https://www.mongodb.com/docs/atlas/build-multi-tenant-arch/) as a reference.
Yes I had one mongodb server in mind which means one database per tenant and all of them on a single mongodb server.
The article says that this setup will allow for a few hundred of tenants(a few hundred databases with the same exact schema 
each containing tenant-specific data).

**Reply by vkarpov15** [reply, marked as answer]
@sadeghhosseini depends on your use case. If you have a relatively small number of tenants (say less than 100) and you're not worried about contention between the tenants, then `useDb()` is fine. `useDb()` is easier to implement and work with, and ends up opening fewer connections to MongoDB.

But `useDb()` means one slow tenant can slow down the other tenants due to [slow trains](https://thecodebarbarian.com/slow-trains-in-mongodb-and-nodejs). That's why, for large multi-tenant setups or setups that are performance sensitive, we recommend a separate connection per tenant.

**Reply by sadeghhosseini** [reply]
@vkarpov15 Thanks for taking the time to answer my question.
Do you recommend the following approach to create a separate connection per tenant?
"creating a connection per tenant and caching the connection objects on a connection map (a JS object or map which keeps connection objects)"

If so is there a mechanism through which I can close a certain connection and remove it from the connection map(cache) after certain amount of inactivity?

Is there any event emitted by mongoose when the MongoDB driver closes all the sockets in a certain connection's pool (on account of them being inactive for certain amount of time)?

MongoClient has "connectionPoolCleared" event which might be what I am looking for, but I couldn't find any decent documentation on it.

**Reply by vkarpov15** [reply]
"creating a connection per tenant and caching the connection objects on a connection map (a JS object or map which keeps connection objects)" <-- yep that works

You can always access the Mongoose connection's MongoClient instance using `connection.client`, for example `connection.client.on('connectionPoolCleared')`. A Mongoose connection always has its own MongoClient instance.
